Expose Node attributes. Support Listener metadata.  Support Cluster M…

jplevyak opened the pull request:
…etadata.

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

jplevyak commented:
This only supports the "local cluster".  I presume we would like to support arbitrary cluster metadata by name.  I am also looking for feedback on the non-metadata metadata (e.g. things in Node) which I am exposing here with the special domain "-".  Again, could do this a different way, looking for feedback, ideas.

kyessenov commented on source/extensions/common/wasm/wasm.cc:
```diff
@@ -983,7 +983,42 @@ const ProtobufWkt::Struct* Context::getMetadataStructProto(MetadataType type,
   case MetadataType::ResponseRoute:
     return getRouteMetadataStructProto(encoder_callbacks_);
   case MetadataType::Node:
+    if (name == "_") {
```
I think it would be better not to supply all the metadata at once, but just have some kind of switch based on keys that only returns individual fields.

jplevyak replied:
You can ask for individual bits of the metadata with the high level API by calling namedMetadataValue("_", "cluster") for example.

jplevyak commented:
I updated this with two different ways to get the upstream cluster name.  Please take a look.

I also changed the way to get "builtin" metadata (e.g. Node attributes) to use the name ".".

jplevyak commented:
Updated wrt latest discussion about how to get the upstream cluster.  PTAL
